### PR TITLE
Add `normalize` matcher

### PR DESCRIPTION
**Pull request by stephannv**
## Why

Rails 7.1 adds `normalizes` API.
- [ActiveRecord::Base::normalizes API.](https://github.com/rails/rails/pull/43945)
- [Docs](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Normalization/ClassMethods.html#method-i-normalizes)

I think would be nice to shoulda-matchers to have a matchers to easily test the new `normalizes` API, similar to `allow_value`/`allow_values` matchers. I'm using on my project and this:
```ruby
it "normalizes email" do
  user = User.new(email: " \n myEmail@here.com \n\n")
  user.normalize_attribute(:email)
  expect(user.email).to eq "myemail@here.com"
end
```

Will become this:
```ruby
it { is_expected.to normalize(:email).from(" \n myEmail@here.com \n\n").to("myemail@here.com") }
```

What do you think about? This PR makes sense? 

Some considerations:
- I'm not sure about the API `normalize(attr)`, `from(value)` and `to(value)`, I'm open to change.
- `normalizes` accepts `apply_to_nil` option, but I think this should be tested in its own spec, like
  ```ruby
  shoulda normalize(:name).from(nil).to("Untitled")
  ```
  Another option would be:
  ```ruby
  shoulda normalize(:name).from(" jane doe ").to("Jane Doe").normalizing_nil_to("Untitled")
  ```
- I had to add Rails 7.1 to Appraisals to be able to test. I'm not sure how to restrict this tests to run only in rails_7_1. Other questions: Do we need to flag this matcher to run only in Rails 7.1 projects? Do we need to add this info to the docs?
- I think this is my first PR to shoulda-matchers, so, feedback are welcome.
- This PR is a work in progress, I opened it to receive feedbacks, so some non-related tests can be broken, or rubocop can be failing.

### CODE REVIEW

**Review by mcmire — COMMENTED** [review]
Hey @stephannv, I think this is a great idea! I would probably suggest that we extract adding Rails 7.1 to another PR as it may cause CI failures. As for whether we should handle the `apply_to_nil` option by adding another method or not, I think we should keep it simple and stick to `from` and `to` like you've done here. 

One thing about this is that we might want to implement a negative version of the matcher. Every matcher can be executed via `should_not`, and that may match or may not match, just as with `should`. We might want to have another failure message in that case. 

Other than that this looks great!

**Comment by stephannv** [comment]
Done @mcmire . Can you take a look again?

I tried to create another PR focused on adding Rails 7.1 to pipeline but I faced some issues, I will give another try later.

**Comment by Sub-Xaero** [comment]
Anything updates on this PR? 
Happy to help out if more needs doing to take it over the line.

**Comment by stephannv** [comment]
> Any updates on this PR? Happy to help out if more needs doing to take it over the line.

@Sub-Xaero I was waiting 7.1 release to update this PR, but I think I don't need to wait.

**Comment by stephannv** [comment]
@Sub-Xaero I remember now, this PR is waiting `shoulda-matchers` to support Rails 7.1. If you want to help, I think you can open a PR with necessary changes to prepare shoulda-matchers for Rails 7.1

**Comment by stephannv** [comment]
Hi @matsales28 , are you the new code owner for this PR? I rebased it with main, I think it's ready for review now. Your feedback will be very much appreciated.

File: lib/shoulda/matchers/active_record/normalize_matcher.rb
Comment (matsales28) [inline lib/shoulda/matchers/active_record/normalize_matcher.rb:133]: To avoid using `public_send` and assigning it to the instance and then performing the normalization, what do you think about using the `normalize_value_for` class method? My initial idea was also to make the subject a class and not an instance, but that would require changing the default subject of the specs, and I don't think that's ideal.

```suggestion
        def normalize_attribute?(subject, attribute)
          subject_class = subject.class
          subject_class.normalize_value_for(attribute, from_value) == to_value
        end
```

**Review by matsales28 — COMMENTED** [review]
Overall it looks good to me. Left some small suggestions but once adjusted we're ready to merge. One thing that could make it even better is when we're dealing with several attributes, it would be awesome if we could check all of them and then provide an error message listing all the attributes that didn't pass the normalization. Right now, the way it's set up, only the first one that doesn't meet the criteria shows up in the error message. We can make this on an next PR.

File: lib/shoulda/matchers/active_record/normalize_matcher.rb
Comment (stephannv) [reply lib/shoulda/matchers/active_record/normalize_matcher.rb:133]: I think it makes sense. I considered using `.normalize_value_for` at first, but I dropped because I wanted to act on the subject itself, but I think it doesn't matter in this case.

**Comment by stephannv** [comment]
> Overall it looks good to me. Left some small suggestions but once adjusted we're ready to merge. One thing that could make it even better is when we're dealing with several attributes, it would be awesome if we could check all of them and then provide an error message listing all the attributes that didn't pass the normalization. Right now, the way it's set up, only the first one that doesn't meet the criteria shows up in the error message. We can make this on an next PR.

Yeah @matsales28 , I think it would be a great improvement. 

Thanks for the feedback, I think I addressed everything. Let me know if I'm missing something.

File: spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb
Comment (matsales28) [inline spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb:87]:
```suggestion
        model = define_model(:User, email: :string)
```

File: spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb
Comment (matsales28) [inline spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb:54]:
```suggestion
        model = define_model(:User, email: :string, name: :string) do
```

File: spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb
Comment (matsales28) [inline spec/unit/shoulda/matchers/active_record/normalize_matcher_spec.rb:35]:
```suggestion
        model = define_model(:User, email: :string) do
```

**Review by matsales28 — CHANGES_REQUESTED** [review]

**Comment by stephannv** [comment]
Sorry for that @matsales28 , I think it's working now.

**Review by matsales28 — APPROVED** [review]

**Pull request merged**
